tensorflow_zero_out/cc/kernels/lsh_matmul.cc

```diff
@@ -0,0 +1,221 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/framework/op_kernel.h"
+
+using namespace tensorflow;
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+
+template <typename Device, typename T, bool USE_CUBLAS>
+
+class MklMatMulOp : public OpKernel {
+ public:
+  explicit MklMatMulOp(OpKernelConstruction* ctx) : OpKernel(ctx) {
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("transpose_a", &transpose_a_));
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("transpose_b", &transpose_b_));
+  }
+
+#define PADDING 4
+#define BISINIT 0
+#define BINDEX 1
+#define BCOUNT 2
+//TODO: fill the values below
+#define BUCKETSIZE 
+#define _L
+#define _numhashes 
+#define _noOfNodes 
+
+  int Bucket_add(int id, int l, int idx, int L, int RangePow, int BUCKETSIZE,
+                 double *lsh_table) {
+
+  int getHashSparse(int* indices, float *values, int length) {
+    int *hashes = new int[_numhashes];
+
+    for (int p = 0; p < _numhashes; p++) {
+        double s = 0;
+        int i = 0;
+        int j = 0;
+        while (i < length & j < _samSize) {
+          if (indices[i] == _indices[p][j]) {
+            float v = values[i];
+            if (_randBits[p][j] >= 0) {
+              s += v;
+            } else {
+              s -= v;
+            }
+            i++;
+            j++;
+          }
+          else if (indices[i] < _indices[p][j]){
+            i++;
+          }
+          else{
+            j++;
+          }
+        }
+        hashes[p] = (s >= 0 ? 0 : 1);
+    }
+    return rawResults;
+  }
+
+    return hashes;
+  }
+
+  int* hashesToIndex(int * hashes) {
+    int * indices = new int[_L];
+    for (int i = 0; i < _L; i++) {
+      unsigned int index = 0;
+      for (int j = 0; j < _K; j++) {
+        if (HashFunction==4) {
+	  unsigned int h = hashes[_K*i + j];
+	  index += h<<(_K-1-j);
+	} 
+      indices[i] = index;
+      }
+    } 
+    return indices;
+  }
+
+  int** retrieveRaw(int *indices) {
+    int ** rawResults = new int*[_L];
+    int count = 0;
+    for (int i = 0; i < _L; i++)
+      rawResults[i] = _bucket[i][indices[i]].getAll();
+    return rawResults;
+  }
+
+  int getCandidates(int **actives) {
+    for(int i = 0; i < _L; i++) {
+      if(actives[i] == NULL) {
+        continue;
+      } 
+      else {
+        for(int j = 0; j < BUCKETSIZE; j++) {
+          int tempID = actives[i][j] - 1;
+          if(tempID >= 0) 
+            counts[tempID] += 1;
+          else
+            break;
+        }
+      }
+    }
+
+    if(counts.size()<1500){
```

Comment:
I think you need to do only one of these approach based on whether we want to do sampling vs top-k